### PR TITLE
[CARBONDATA-3337] Implemented Hadoop RPC framework for index server

### PR DESCRIPTION
**Pull request by kunal642**
Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [ ] Any interfaces changed?
 
 - [ ] Any backward compatibility impacted?
 
 - [ ] Document update required?

 - [ ] Testing done
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
        - How it is tested? Please attach test report.
        - Is it a performance related change? Please attach the performance test report.
        - Any additional information to help reviewers in testing this change.
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2838/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/11098/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/3069/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2839/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2840/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/3071/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/11100/

File: integration/spark2/src/main/scala/org/apache/carbondata/indexserver/IndexServer.scala [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+        (null, "local")
+      }
+      createCarbonSession(storePath, masterURL)
+      LOGGER.info("Starting Index Cache Server")
+      val conf = new Configuration()
+      val server: RPC.Server = new RPC.Builder(conf).setInstance(this)
+        .setBindAddress(serverIp)
+        .setPort(serverPort)
+        .setProtocol(classOf[ServerInterface]).build
+      server.start()
```
Comment (qiuchenjian) [inline integration/spark2/src/main/scala/org/apache/carbondata/indexserver/IndexServer.scala]: Does start function have return code to judge the status of startting?

Comment (kunal642) [reply integration/spark2/src/main/scala/org/apache/carbondata/indexserver/IndexServer.scala]: No, return type is void

File: core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java [hunk 2]
```diff
@@ -2104,4 +2104,17 @@ private CarbonCommonConstants() {
    */
   public static final String CARBON_QUERY_DATAMAP_BLOOM_CACHE_SIZE_DEFAULT_VAL = "512";
 
-}
+  public static final String CARBON_INDEX_SERVER_POLICY = "carbon.index.server.policy";
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java]: please write description for this config in comment, like in what scenario this config is used, what valid values are, etc

File: core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java [hunk 3]
```diff
@@ -2104,4 +2104,17 @@ private CarbonCommonConstants() {
...
+
+  public static final String CARBON_DISTRIBUTED_PRUNE_POLICY = "distributed";
+
+  public static final String CARBON_EMBEDDED_PRUNE_POLICY = "embedded";
+
+  public static final String CARBON_DRIVER_PRUNE_POLICY = "driver";
+
+  public static final String CARBON_INDEX_SERVER_IP = "carbon.index.server.ip";
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java]: please write comment for all new configuration

File: core/src/main/java/org/apache/carbondata/core/util/CarbonProperties.java [hunk 4]
```diff
@@ -1601,4 +1601,47 @@ private void validateDetailQueryBatchSize() {
       }
     }
   }
-}
+
+  public String getIndexServerPolicy() {
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/util/CarbonProperties.java]: please add comment for all public function

File: core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java [hunk 5]
```diff
@@ -2104,4 +2104,17 @@ private CarbonCommonConstants() {
    */
   public static final String CARBON_QUERY_DATAMAP_BLOOM_CACHE_SIZE_DEFAULT_VAL = "512";
 
-}
+  public static final String CARBON_INDEX_SERVER_POLICY = "carbon.index.server.policy";
+
+  public static final String CARBON_DISTRIBUTED_PRUNE_POLICY = "distributed";
+
+  public static final String CARBON_EMBEDDED_PRUNE_POLICY = "embedded";
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java]: create a "CARBON_INDEX_SERVER_POLICY_DEFAULT" constant

File: core/src/main/java/org/apache/carbondata/core/util/CarbonProperties.java [hunk 6]
```diff
@@ -1601,4 +1601,47 @@ private void validateDetailQueryBatchSize() {
...
+    if (indexServerPolicy.equalsIgnoreCase(CarbonCommonConstants.CARBON_DISTRIBUTED_PRUNE_POLICY)
+        || indexServerPolicy.equalsIgnoreCase(CarbonCommonConstants.CARBON_EMBEDDED_PRUNE_POLICY)
+        || indexServerPolicy.equalsIgnoreCase(CarbonCommonConstants.CARBON_DRIVER_PRUNE_POLICY)) {
+      LOGGER.info("The prune policy set for pruning is: " + indexServerPolicy);
+      return indexServerPolicy;
+    } else {
+      LOGGER.info("The configured value for " + CarbonCommonConstants.CARBON_INDEX_SERVER_POLICY
+          + " is not valid, therefore taking " + CarbonCommonConstants.CARBON_EMBEDDED_PRUNE_POLICY
+          + " as the index server prune policy");
+      return CarbonCommonConstants.CARBON_EMBEDDED_PRUNE_POLICY;
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/util/CarbonProperties.java]: use CARBON_INDEX_SERVER_POLICY_DEFAULT

File: integration/spark2/src/main/scala/org/apache/carbondata/indexserver/DistributedDataMapJob.scala [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+import java.util
+
+import scala.collection.JavaConverters._
+
+import org.apache.carbondata.core.datamap.{AbstractDataMapJob, DistributableDataMapFormat}
+import org.apache.carbondata.core.indexstore.ExtendedBlocklet
+import org.apache.carbondata.core.scan.filter.resolver.FilterResolverIntf
+
+class DistributedDataMapJob extends AbstractDataMapJob {
```
Comment (jackylk) [inline integration/spark2/src/main/scala/org/apache/carbondata/indexserver/DistributedDataMapJob.scala]: Why is this class in spark2 integration module? Can't put in carbondata-core?

File: integration/spark2/src/main/scala/org/apache/carbondata/indexserver/DistributedPruneRDD.scala [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+import org.apache.spark.{Partition, TaskContext}
+import org.apache.spark.sql.SparkSession
+
+import org.apache.carbondata.core.datamap.DistributableDataMapFormat
+import org.apache.carbondata.core.indexstore.ExtendedBlocklet
+import org.apache.carbondata.core.scan.filter.resolver.FilterResolverIntf
+import org.apache.carbondata.spark.rdd.CarbonRDD
+
+
+class DistributedPruneRDD(@transient private val ss: SparkSession,
```
Comment (jackylk) [inline integration/spark2/src/main/scala/org/apache/carbondata/indexserver/DistributedPruneRDD.scala]: write comment please

File: integration/spark2/src/main/scala/org/apache/carbondata/indexserver/DistributedPruneRDD.scala [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+    resolverIntf: FilterResolverIntf)
+  extends CarbonRDD[(String, ExtendedBlocklet)](ss, Nil) {
+
+  override protected def internalGetPartitions: Array[Partition] = {
+    Array()
+  }
+
+  override def internalCompute(split: Partition,
+      context: TaskContext): Iterator[(String, ExtendedBlocklet)] = {
+    Nil.iterator
```
Comment (jackylk) [inline integration/spark2/src/main/scala/org/apache/carbondata/indexserver/DistributedPruneRDD.scala]: No implementation?

File: integration/spark2/src/main/scala/org/apache/carbondata/indexserver/IndexServer.scala [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+  }
+
+  private def createCarbonSession(storePath: String, masterURL: String): SparkSession = {
+    import org.apache.spark.sql.CarbonSession._
+    val master = System.getProperty("spark.master.url", masterURL)
+    if (master == null) {
+      throw new InvalidConfigurationException("Spark master URL is not set.")
+    }
+    val spark = SparkSession
+      .builder().config(new SparkConf())
```
Comment (jackylk) [inline integration/spark2/src/main/scala/org/apache/carbondata/indexserver/IndexServer.scala]: move config together

File: integration/spark2/src/main/scala/org/apache/carbondata/indexserver/IndexServer.scala [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+
+@ProtocolInfo(protocolName = "Server", protocolVersion = 1)
+trait ServerInterface {
+  def getSplits(request: DistributableDataMapFormat): Array[(String, ExtendedBlocklet)]
+  def invalidateCache(tableInfo: TableInfo): Unit
+  def showCache()
+}
+
+object IndexServer extends ServerInterface {
+  val prunePolicy: String = CarbonProperties.getInstance().getIndexServerPolicy
```
Comment (jackylk) [inline integration/spark2/src/main/scala/org/apache/carbondata/indexserver/IndexServer.scala]: use these variables in main directly, no need to declare local variables

File: integration/spark2/src/main/scala/org/apache/carbondata/indexserver/IndexServer.scala [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+      .enableHiveSupport()
+      .config("spark.dynamicAllocation.enabled", "false")
+      .getOrCreateCarbonSession(storePath)
+    SparkSession.setActiveSession(spark)
+    SparkSession.setDefaultSession(spark)
+    spark
+  }
+
+  def getClient: ServerInterface = {
+    import org.apache.hadoop.ipc.RPC
```
Comment (jackylk) [inline integration/spark2/src/main/scala/org/apache/carbondata/indexserver/IndexServer.scala]: move to head

File: integration/spark2/src/main/scala/org/apache/carbondata/indexserver/IndexServer.scala [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+
+object IndexServer extends ServerInterface {
+  val prunePolicy: String = CarbonProperties.getInstance().getIndexServerPolicy
+
+  private val LOGGER = LogServiceFactory.getLogService(this.getClass.getName)
+  private val serverIp: String = CarbonProperties.getInstance().getIndexServerIP
+
+  private val serverPort: Int = CarbonProperties.getInstance().getIndexServerPort
+
+  def getSplits(request: DistributableDataMapFormat): Array[(String, ExtendedBlocklet)] = {
```
Comment (jackylk) [inline integration/spark2/src/main/scala/org/apache/carbondata/indexserver/IndexServer.scala]: no `override`?

**Comment by kunal642** [comment]
@jackylk This is the old PR..Can you review the new PR #3177  instead of this. I will handle the comments of this PR in 3177

**Pull request closed without merging**